python/ql/consistency-queries/DataFlowConsistency.ql

```diff
@@ -120,6 +140,9 @@ private module Input implements InputSig<PythonDataFlow> {
       arg = other.getArgument(any(ArgumentPosition p | p.isSelf())) and
       other.getNode().getScope() = f
     )
+    or
+    arg = call.getArgument(_) and
+    arg instanceof CaptureArgumentNode
```

Comment:
Please add a comment as to why this is needed

Reply:
Done

Reply:
Thanks. I don't quite understand why it isn't covered by

```codeql
    exists(DataFlowCall other, CallNode cfgCall | other != call |
      call.getNode() = cfgCall and
      other.getNode() = cfgCall and
      isArgumentNode(arg, call, _) and
      isArgumentNode(arg, other, _)
    )
```

I've tried to come up with an example where the same ("lambda self") argument would be used for multiple calls:

```py
def outer():
    var = 42
    def foo():
        print("foo", var)
    def bar():
        print("bar", var)
    var = 43
    f = foo if <cond> else bar
    f()
```

In the `f()` call we would either call foo or bar, who would have captured `var`, and with the changes in this PR, we should be able to correctly identify that the value printed is `43`.

There will be TWO `DataFlowCall` for `f()`, one that targets `foo` and one that targets `bar`.

There will be ONE `CaptureArgumentNode` which is based on the cfg-node for .getFunction, meaning `f` in `f()`.

From the way I see it, that should 100% be covered by the exception above.


### Interplay with self argument

In the char-pred of `ExtractedArgumentNode` we have:

https://github.com/github/codeql/blob/f96c52ed3bc117f7c51edc1cf5f1a3ed83d7a6e3/python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPublic.qll#L310-L311

So RIGHT NOW it's the case that `forall(CaptureArgumentNode arg | | arg instanceof ExtractedArgumentNode)` holds.

So maybe this is consistency exception is needed due to the cross-talk I feared here: https://github.com/github/codeql/pull/14944#discussion_r1426887665 ?

Reply:
The exclusion case you reference require the two calls to be the same `CallNode`, I think?

Reply:
You were right, it does not seem to be needed. It is removed in https://github.com/github/codeql/pull/14944/commits/d6544cc5505f9cac8f4f2477e035cf2b9755f84d without any consistency failures appearing..

Reply:
Nice! :+1: